include/swift/AST/LocalizationFormat.h

```diff
@@ -43,7 +131,29 @@ class LocalizationProducer {
 class YAMLLocalizationProducer final : public LocalizationProducer {
 public:
   std::vector<std::string> diagnostics;
-  explicit YAMLLocalizationProducer(std::string locale, std::string path);
+  explicit YAMLLocalizationProducer(llvm::StringRef filePath);
+  llvm::StringRef getMessageOr(swift::DiagID id,
+                               llvm::StringRef defaultMessage) const override;
+};
+
+class SerializedLocalizationProducer final : public LocalizationProducer {
+  using SerializedLocalizationTable =
+      llvm::OnDiskIterableChainedHashTable<LocalizationLookupInfo>;
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Buffer;
+  llvm::BitstreamCursor Cursor;
+  std::unique_ptr<SerializedLocalizationTable> SerializedTable;
+
+public:
+  explicit SerializedLocalizationProducer(llvm::StringRef filePath)
+      : Buffer(llvm::MemoryBuffer::getFileOrSTDIN(filePath)),
+        Cursor(*Buffer.get()) {
+    auto base = Cursor.getBitcodeBytes();
+    const uint8_t *B = base.data();
+    uint32_t offset = endian::readNext<uint32_t, little, unaligned>(B);
```

Comment:
I think instead of even using `Cursor` here, you could use `getBufferStart()` because these "offset" bytes are always stored at `0`, so `Cursor` could be removed completely.